File: leetcode/editor/cn/[283]移动零.go
//给定一个数组 nums，编写一个函数将所有 0 移动到数组的末尾，同时保持非零元素的相对顺序。 
//
// 示例: 
//
// 输入: [0,1,0,3,12]
//输出: [1,3,12,0,0] 
//
// 说明: 
//
// 
// 必须在原数组上操作，不能拷贝额外的数组。 
// 尽量减少操作次数。 
// 
// Related Topics 数组 双指针 
// 👍 828 👎 0
package main

import "fmt"

//leetcode submit region begin(Prohibit modification and deletion)
func moveZeroes(nums []int)  {
	begin := 0
	for i := 0; i < len(nums); i++ {
		if nums[i] != 0 {
			nums[i], nums[begin] = nums[begin], nums[i]
			begin += 1
		}
	}
}
//leetcode submit region end(Prohibit modification and deletion)

func main() {
	nums := []int {0, 1, 0, 3, 12}
	except := []int {1, 3, 12, 0, 0}
	moveZeroes(nums)

	for i := 0; i < len(nums); i++ {
		if nums[i] != except[i] {
			fmt.Println("err", nums, except)
			break
		}
	}
}

File: leetcode/editor/cn/[283]移动零.py
# 给定一个数组 nums，编写一个函数将所有 0 移动到数组的末尾，同时保持非零元素的相对顺序。 
# 
#  请注意 ，必须在不复制数组的情况下原地对数组进行操作。 
# 
#  
# 
#  示例 1: 
# 
#  
# 输入: nums = [0,1,0,3,12]
# 输出: [1,3,12,0,0]
#  
# 
#  示例 2: 
# 
#  
# 输入: nums = [0]
# 输出: [0] 
# 
#  
# 
#  提示: 
#  
# 
#  
#  1 <= nums.length <= 10⁴ 
#  -2³¹ <= nums[i] <= 2³¹ - 1 
#  
# 
#  
# 
#  进阶：你能尽量减少完成的操作次数吗？ 
#  Related Topics 数组 双指针 👍 1532 👎 0


# leetcode submit region begin(Prohibit modification and deletion)
from typing import List


class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        双指针，左指针指向当前非0位置，右指针向右遍历，如果当前数非0，则将其移动到左指针位置
        """
        l, r = 0, 0
        while r < len(nums):
            if nums[r] != 0:
                nums[l], nums[r] = nums[r], nums[l]
                l = l + 1
            r = r + 1


# leetcode submit region end(Prohibit modification and deletion)
if __name__ == '__main__':
    nums = [0, 1, 0, 3, 12]
    Solution().moveZeroes(nums)
    assert nums == [1, 3, 12, 0, 0]

    nums = [0]
    Solution().moveZeroes(nums)
    assert nums == [0]
